File: DataStructures/php/LinkedList/LinkedList.php
<?php

class Node
{
    public $data;
    public ?Node $next;
    public ?Node $prev;

    public function __construct($data)
    {
        $this->data = $data;
        $this->next = null;
        $this->prev = null;
    }
}

class LinkedList
{
    public Node $head;
    public Node $tail;
    public int $length = 0;

    public function __construct()
    {
        $this->head = new Node(null);
        $this->tail = new Node(null);

        $this->head->next = $this->tail;
        $this->tail->prev = $this->head;
    }

    public function append(Node $node)
    {
        $node->prev = $this->tail->prev;
        $node->prev->next = $node;
        $this->tail->prev = $node;
        $node->next = $this->tail;

        $this->length++;
    }

    public function push(Node $node)
    {
        $node->next = $this->head->next;
        $this->head->next = $node;
        $node->prev = $this->head;
        $node->next->prev = $node;

        $this->length++;
    }

    public function pop()
    {
        if ($this->length > 0) {
            $this->tail->prev->prev->next = $this->tail;
            $this->tail->prev = $this->tail->prev->prev;
            $this->length--;
        }
    }

    public function shift()
    {
        if($this->length > 0)
        {
            $this->head->next = $this->head->next->next;
            $this->head->next->prev = $this->head;
            $this->lenght --;
        }
    }

    public function insert(int $index, Node $node)
    {
    }

    public function remove(int $index)
    {
        if($index > $this->length)
        {
            return "index out of range";
        }

        $target = $this->getNode($index);

        $target->next->prev = $target->prev;
        $target->prev->next = $target->next;
        $this->length --;
    }

    public function get(int $index)
    {
        return $this->getNode($index)->data;
    }

    private function getNode(int $index)
    {
        if ($index > $this->length) {
            return "index out of range";
        }

        $item = $this->head;
        for ($i = 0; $i < $index; $i++) {
            $item = $item->next;
        }

        return $item;
    }

    public function length()
    {
        return $this->length;
    }

    public function isEmpty()
    {
        return $this->length() === 0;
    }
}

$list = new LinkedList();
echo "list lenght 0 = " . $list->length();
echo PHP_EOL;

$list->append(new Node(20));
echo "list lenght 1 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 20 = " . $list->get(1);
echo PHP_EOL;

$list->append(new Node(30));
echo "list lenght 2 = " . $list->length();
echo PHP_EOL;
echo "list second Node data 30 = " . $list->get(2);
echo PHP_EOL;

$list->pop();
echo "list lenght 1 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 20 = " . $list->get(1);
echo PHP_EOL;
echo "list second Node not exist = " . $list->get(2);
echo PHP_EOL;

$list->push(new Node('amir'));
echo "list lenght 2 = " . $list->length();
echo PHP_EOL;
echo "list first Node data amir = " . $list->get(1);
echo PHP_EOL;
echo "list second Node 20 = " . $list->get(2);
echo PHP_EOL;

$list->push(new Node(696));
echo "list lenght 3 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 696 = " . $list->get(1);
echo PHP_EOL;
echo "list second Node amir = " . $list->get(2);
echo PHP_EOL;
echo "list third Node 20 = " . $list->get(3);
echo PHP_EOL;

$list->remove(2);
echo "list lenght 2 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 696 = " . $list->get(1);
echo PHP_EOL;
echo "list second Node 20 = " . $list->get(2);
echo PHP_EOL;

$list->append('append');
echo "list lenght 3 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 696 = " . $list->get(1);
echo PHP_EOL;
echo "list second Node 20 = " . $list->get(2);
echo PHP_EOL;
echo "list third Node append = " . $list->get(3);
echo PHP_EOL;

$list->shift();
echo "list lenght 2 = " . $list->length();
echo PHP_EOL;
echo "list first Node data 20 = " . $list->get(1);
echo PHP_EOL;
echo "list second Node append = " . $list->get(2);

File: DataStructures/python/LinkeList/LinkedList.py
from contextlib import nullcontext
from operator import length_hint


class Node:

    def __init__(self, data=None) -> None:
        self.data = data
        self.next = None
        self.prev = None

class LinkedList:

    def __init__(self) -> None:
        self.head = Node()
        self.tail = Node()

        self.head.next = self.tail
        self.tail.prev = self.head

        self.length = 0

    def append(self, node):
        node.next = self.tail
        node.prev = self.tail.prev
        self.tail.prev = node
        node.prev.next = node
        self.length += 1

    def push(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next = node
        node.next.prev = node
        self.length += 1

    def pop(self):
        if self.length < 1:
            return None
        self.tail.prev.prev.next = self.tail
        self.tail.prev = self.tail.prev.prev
        self.length -= 1

    def shift(self):
        if self.length < 1:
            return None

        self.head.next.next.prev = self.head
        self.head.next = self.head.next.next

        self.length -= 1

    def insert(self, index, node):

        if index == 1:
            return self.push(node)
        
        if index == self.length:
            return self.append(node)

        if 0 >= index > self.length:
            return None

        target = self.head
        for _ in range(index):
            target = target.next
        
        node.next = target
        node.prev = target.prev
        target.prev = node
        node.prev.next = node

        self.length += 1


    def get(self, index):
        if index <= 0 or index > self.length:
            return None

        target = self.head

        for _ in range(index):
            target = target.next
        
        return target.data

    def len(self):
        return self.length

    def isEmpty(self):
        return self.length == 0
    
    def __str__(self) -> str:
        
        node = self.head.next
        output = "["
        for _ in range(self.length):
            sep = '' if _ == self.length -1 else ", "
            output += node.data + sep
            node = node.next
        output += ']'

        return output

linkeList = LinkedList()

assert linkeList.length == 0
linkeList.append(Node('append'))
assert linkeList.length == 1
print('length =', linkeList.len())
assert linkeList.get(1) == "append"
print('get first =', linkeList.get(1))
print(linkeList)
print('*' * 100)

assert linkeList.length == 1
linkeList.push(Node('push'))
assert linkeList.length == 2
print('length =', linkeList.len())
assert linkeList.get(1) == "push"
assert linkeList.get(2) == "append"
print('get first =', linkeList.get(1))
print(linkeList)
print('*' * 100)

linkeList.pop()
assert linkeList.length == 1
assert linkeList.get(1) == "push"
print('length =', linkeList.len())
print(linkeList)
print('*' * 100)

linkeList.shift()
assert linkeList.length == 0
print('length =', linkeList.len())
print(linkeList)
print('*' * 100)

linkeList.insert(1, Node('insert 1'))
assert linkeList.length == 1
assert linkeList.get(1) == "insert 1"
linkeList.insert(1, Node('insert 2'))
print('length =', linkeList.len())
assert linkeList.get(1) == "insert 2"
assert linkeList.get(2) == "insert 1"
assert linkeList.length == 2
print(linkeList)
print('*' * 100)